File: ios/Athkar/Domain/Coordinates.swift
import Foundation

/// A geographic position in signed decimal degrees. North and east are positive.
struct Coordinates: Equatable, Codable, Sendable {
    let latitude: Double
    let longitude: Double

    /// Returns nil rather than trapping: these values often arrive from storage or a sensor, and a
    /// bad pair should send the user back to the picker, not crash the app.
    init?(latitude: Double, longitude: Double) {
        guard latitude.isFinite, longitude.isFinite,
              (-90.0...90.0).contains(latitude),
              (-180.0...180.0).contains(longitude)
        else { return nil }
        self.latitude = latitude
        self.longitude = longitude
    }
}

File: android/core/src/main/kotlin/com/athkar/core/prayer/Coordinates.kt
package com.athkar.core.prayer

/**
 * A geographic position in signed decimal degrees. North and east are positive.
 *
 * @property latitude degrees, `-90..90`.
 * @property longitude degrees, `-180..180`.
 */
data class Coordinates(val latitude: Double, val longitude: Double) {
    init {
        require(!latitude.isNaN() && latitude in -90.0..90.0) {
            "latitude must be between -90 and 90, was $latitude"
        }
        require(!longitude.isNaN() && longitude in -180.0..180.0) {
            "longitude must be between -180 and 180, was $longitude"
        }
    }
}
